### PR TITLE
[ENH] Implemented Tracking differentiator-based multiview dilated characteristics (TD-MVDC) Classifier

### LINKED ISSUE #2081: [ENH] Implement the feature-based TD-MVDC classification algorithm
### Describe the feature or idea you want to propose

Tracking differentiator-based multiview dilated characteristics (TD-MVDC) is a new feature-based TSC algorithm.

The introduction of a tracking differentiator combined with dilation mapping as a preprocessor into the feature-based TSC method is proposed to improve feature diversity of TSFresh efficiently.
Ensemble feature selection based on filter feature selectors with different store ratios is designed to generate multiview features to enhance feature stability quickly.
Linear classifiers and hard voting to fastly classify and integrate multiview features to increase classification performance robustly.

The paper title is " Tracking Differentiator-based Multiview Dilated Characteristics for Time Series Classification" and has been peer-reviewed and presented at the INDIN24 conference.
The paper file and Python code can be obtained below:
https://github.com/CCHe64/TD-MVDC
Preprinted version:
https://github.com/CCHe64/TD-MVDC/blob/main/Tracking%20Differentiator-based%20Multiview%20Dilated%20Characteristics%20for%20Time%20Series%20Classification.pdf
TD-MVDC is implemented through functions in sklearn and aeon.
Python source code is here:
https://github.com/CCHe64/TD-MVDC/blob/main/Tracking%20Differentiator-based%20Multiview%20Dilated%20Characteristics.ipynb

### Describe your proposed solution

Firstly, in aeon implement a new feature extraction transformation function using TSFreshEigenExtractor in transformations. collection.feature_based.
Then implement a new classification function in aeon. classification.feature_based.

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lucifer4073**
#### Reference Issues/PRs

Closes #2081. 

#### What does this implement/fix? Explain your changes.

This PR implements the TDMVDC classifier.

#### Does your contribution introduce a new dependency? If yes, which one?

No

#### Any other comments?

Currently using the TSFreshRelevant feature extractor for feature extraction. I have tried implementing a similar module (for feature extraction) dedicated for tdmvdc itself (doesn't give great results).  

Accuracy as per paper: 81.14 per cent
Obtained accuracy: 77.71 per cent

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you after the PR has been merged.
- [X] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [ ] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

**Comment by lucifer4073** [comment]
Hi @MatthewMiddlehurst @TonyBagnall @CCHe64 ,

This is the current working implementation of TDMVDC. Further additions would be needed in test coverage and multithreading, which I would like to put up in another PR, once this is merged.

Let me know you reviews.

Thanks!

File: aeon/classification/feature_based/_tdmvdc.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+)
+
+__all__ = ["TDMVDCClassifier"]
+
+
+class TDMVDCClassifier(BaseClassifier):
+    """
+    Tracking Differentiator-based Multiview Dilated Characteristics.
+
+    for Time Series Classification.
```
Comment (MatthewMiddlehurst) [inline aeon/classification/feature_based/_tdmvdc.py:22]: This needs an extended description. See the documentation developer guide.

File: aeon/classification/feature_based/_tdmvdc.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
+import numpy as np
```
Comment (MatthewMiddlehurst) [inline aeon/classification/feature_based/_tdmvdc.py:1]: Include documentation at the top of the file.

File: aeon/classification/feature_based/_tdmvdc.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    }
+
+    def __init__(
+        self,
+        default_fc_parameters="efficient",
+        k1=2,
+        k2=2,
+        feature_store_ratios=None,
+        n_jobs=1,
+    ):  # Remove **kwargs from parameters
```
Comment (MatthewMiddlehurst) [inline aeon/classification/feature_based/_tdmvdc.py:61]: Why is this comment here?

File: aeon/classification/feature_based/_tdmvdc.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+            feature_store_ratios
+            if feature_store_ratios is not None
+            else [0.1, 0.2, 0.3, 0.4, 0.5]
+        )
+        self.feature_store_ratios = feature_store_ratios
+        self.n_jobs = n_jobs
+
+        super().__init__()
+
+    def _fit(self, trainSignalX, trainY):
```
Comment (MatthewMiddlehurst) [inline aeon/classification/feature_based/_tdmvdc.py:75]: Please just use X and y like other estimators.

File: aeon/classification/feature_based/_tdmvdc.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+                ]
+            )
+            bestIndex_ = np.argsort(self.scoreRFS)[::-1][
+                0 : int(len(self.scoreRFS) * ratio_)
+            ]  # The feature indexes of the top scores
+            clf.fit(np.hstack((trainRX, trainFX, trainSX))[:, bestIndex_], trainY)
+            self.clfList.append(clf)  # Saving the trained classifier
+        return self
+
+    def _predict(self, testSignalX):
```
Comment (MatthewMiddlehurst) [inline aeon/classification/feature_based/_tdmvdc.py:185]: Same here.

File: aeon/classification/feature_based/tests/test_tdmvdc.py
Comment (MatthewMiddlehurst) [inline aeon/classification/feature_based/tests/test_tdmvdc.py:1]: This is not sufficient testing.

File: aeon/transformations/collection/feature_based/_tdmvdc_extractor.py
Comment (MatthewMiddlehurst) [inline aeon/transformations/collection/feature_based/_tdmvdc_extractor.py:1]: This module is for transformer classes only. If this does not fit as a transformer, it can be private functions in the classifier file.

**Review by MatthewMiddlehurst — CHANGES_REQUESTED** [review]
I have had a quick look, not really at any of the estimator code itself. Still a few comments for now.

Do you have an implementation or publication results you can compare against your implementation against? Was this code inspired by any other implementations?

**Review by MatthewMiddlehurst — COMMENTED** [review]
It is unlikely we will merge this if it does not achieve similar results to the original.

**Comment by lucifer4073** [comment]
> It is unlikely we will merge this if it does not achieve similar results to the original.

Hi,

I have tested this of it on the load_arrow head dataset and it achieves the same accuracy as the paper.

**Comment by MatthewMiddlehurst** [comment]
Could you write a test or provide some example code and images showing that? Using more than one dataset would also be good.

**Comment by lucifer4073** [comment]
example code has been added for load_arrow_head. Will add some test cases for the same.

**Comment by CCHe64** [comment]
Thank you very much for your @TonyBagnall @MatthewMiddlehurst review and implementation of @lucifer4073 .
I reviewed the code and performed classification experiments on the 85 UCR datasets.
In the original paper, the average accuracy on 85 datasets is 84.13%.
The Aeon implementation of TD-MVDC achieves an average accuracy of 84.07% on 85 UCR datasets.
In addition, the number of win-lose datasets for both is very close.
Therefore, I am happy with the Aeon version implementation.
I recommend setting the default parameter of "feature_store_ratios" to [0.1, 0.2, 0.3, 0.4, 0.5] as in the original paper instead of None.
I had no issues after making this change.

**Comment by MatthewMiddlehurst** [comment]
Great to hear it performs as expected. Thanks for running it. The code and testing still needs some improvements before this can be merged, however.

**Comment by lucifer-reborn473** [comment]
Hi @CCHe64 @MatthewMiddlehurst ,

i have made the given changes. As far the tests are concerned I checked the compatibility for certain datasets. Let me know if this needs to be extended.

**Comment by lucifer4073** [comment]
Hi @CCHe64 @MatthewMiddlehurst ,

I have made the requested changes.
As far the testing is concerned, I have only checked for a handful of datasets (accounting the execution time). 
Let me know if any improvements are needed.

**Comment by CCHe64** [comment]
> Hi @CCHe64 @MatthewMiddlehurst ,
> 
> I have made the requested changes. As far the testing is concerned, I have only checked for a handful of datasets (accounting the execution time). Let me know if any improvements are needed.

I am happy with this implementation, thanks for your work.

File: aeon/testing/testing_config.py [hunk 6]
```diff
@@ -49,6 +49,7 @@
     # needs investigation
     "SASTClassifier": ["check_fit_deterministic"],
     "RSASTClassifier": ["check_fit_deterministic"],
+    "TDMVDCClassifier": ["check_constructor"],
```
Comment (MatthewMiddlehurst) [inline aeon/testing/testing_config.py:52]: why?

File: aeon/classification/feature_based/tests/test_tdmvdc.py [hunk 7]
```diff
@@ -15,3 +23,44 @@ def test_tdmvdc_classifier():
...
+        testSignalX, testY = load_arrow_head("TEST")
+    elif dataset_name == "GunPoint":
+        trainSignalX, trainY = load_gunpoint("TRAIN")
+        testSignalX, testY = load_gunpoint("TEST")
+    elif dataset_name == "ItalyPowerDemand":
+        trainSignalX, trainY = load_italy_power_demand("TRAIN")
+        testSignalX, testY = load_italy_power_demand("TEST")
+    else:
+        trainSignalX, trainY = load_classification(dataset_name, "TRAIN")
+        testSignalX, testY = load_classification(dataset_name, "TEST")
```
Comment (MatthewMiddlehurst) [inline aeon/classification/feature_based/tests/test_tdmvdc.py:56]: I would remove these, dont need so many.

File: examples/classification/feature_based.ipynb [hunk 8]
```diff
@@ -419,20 +344,25 @@
...
-     "data": {
-      "text/plain": [
-       "(112, 4)"
-      ]
-     },
-     "execution_count": 3,
-     "metadata": {},
-     "output_type": "execute_result"
+     "ename": "ValueError",
+     "evalue": "Unknown estimator name TDMVDC. For a list of valid names and allowed aliases, see NAME_ALIASES in aeon/benchmarking/results_loaders.py.",
```
Comment (MatthewMiddlehurst) [inline examples/classification/feature_based.ipynb:354]: The notebook now raises an error.

File: aeon/classification/feature_based/_tdmvdc.py [hunk 9]
```diff
@@ -0,0 +1,447 @@
+"""Tracking Differentiator-based Multiview Dilated Characteristics (TDMVDC).
+
+Enhances the discriminative power of time series data by applying a
+combination of tracking differentiator-based transformations and multiview
+dilated feature extraction. Leverages advanced feature engineering techniques
+to capture both the original and differential characteristics of time series at
+multiple temporal resolutions. Integrates feature selection and ensemble
+learning to improve classification accuracy and robustness across diverse time
+series datasets.
```
Comment (MatthewMiddlehurst) [inline aeon/classification/feature_based/_tdmvdc.py:9]: Seems unnecessary if the content is duplicated? The docstring description is already rather large.

File: aeon/transformations/collection/feature_based/_tdmvdc_extractor.py
Comment (MatthewMiddlehurst) [reply aeon/transformations/collection/feature_based/_tdmvdc_extractor.py:1]: This does not seem resolved as the transformer file is still there and the functions are public?

Comment (TonyBagnall) [reply aeon/transformations/collection/feature_based/_tdmvdc_extractor.py:1]: I agree it should be a private function in the classifier file, unless it is made into an aeon BaseCollectionTransformer

File: aeon/classification/feature_based/tests/test_tdmvdc.py (hunk 7 above)
Comment (lucifer4073) [reply aeon/classification/feature_based/tests/test_tdmvdc.py:56]: Works. Then the test would check for just 3 datasets (ArrowHead, GunPoint, ItalyPowerDemand).

File: aeon/classification/feature_based/_tdmvdc.py (hunk 9 above)
Comment (lucifer4073) [reply aeon/classification/feature_based/_tdmvdc.py:9]: Sure, will reduce it.

**Review by MatthewMiddlehurst — APPROVED** [review]

**Comment by lucifer4073** [comment]
Hi @MatthewMiddlehurst ,

I believe we can merge this now.
